### PR TITLE
Documentation: document upgrading to v3.1

### PR DESCRIPTION
**Pull request by gyuho**

### CODE REVIEW

**Comment by gyuho** [comment]
@xiang90 PTAL.

File: Documentation/upgrades/upgrade_3_1.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+Before beginning, [backup the etcd data](../op-guide/maintenance.md#snapshot-backup). Should something go wrong with the upgrade, it is possible to use this backup to [downgrade](#downgrade) back to existing etcd version.
+
+#### Mixed Versions
+
+While upgrading, an etcd cluster supports mixed versions of etcd members, and operates with the protocol of the lowest common version. The cluster is only considered upgraded once all of its members are upgraded to version 3.1. Internally, etcd members negotiate with each other to determine the overall cluster version, which controls the reported version and the supported features.
+
+#### Limitations
+
+It might take several seconds for the newly upgraded member to catch up with the existing cluster when the total data size is larger than 700MB. Please expect intermittent downtime and simulate recovery scenarios beforehand.
```
Comment (xiang90) [inline Documentation/upgrades/upgrade_3_1.md]: This does not seem like a limitation anymore. etcd works as expected. people should expect that transferring data might take some time. I think we should change the previous limitation to something like:

If you have large v2 data, then...

If you only have data in v3 backend, you are not subject to this limitation.

Comment (gyuho) [inline Documentation/upgrades/upgrade_3_1.md]: @xiang90 Right. Just fixed.

File: Documentation/upgrades/upgrade_3_1.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+While upgrading, an etcd cluster supports mixed versions of etcd members, and operates with the protocol of the lowest common version. The cluster is only considered upgraded once all of its members are upgraded to version 3.1. Internally, etcd members negotiate with each other to determine the overall cluster version, which controls the reported version and the supported features.
+
+#### Limitations
+
+If you have large v2 data, it might take up to 2 minutes for the newly upgraded member to catch up with the existing cluster when the total data size is larger than 50MB. Check the size of a recent  snapshot to estimate  the total data size. In other words, it is safest to wait for 2 minutes between upgrading each member.
+
+For a much larger total data size, 100MB or more , this one-time process might take even more time. Administrators of very large etcd clusters of this magnitude can feel free to contact the [etcd team][etcd-contact] before upgrading, and we’ll be happy to provide advice on the procedure.
+
+If you only have data in v3 backend, you are not subject to this limitation.
```
Comment (xiang90) [inline Documentation/upgrades/upgrade_3_1.md]: move this as first sentence?

**Comment by xiang90** [comment]
LGTM. defer to @heyitsanthony

File: Documentation/upgrades/upgrade_3_1.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+2017-01-17 09:36:38.406295 W | etcdserver: member fd32987dcd0511e0 has a higher version 3.1.0
+2017-01-17 09:36:42.407695 W | etcdserver: the local etcd version 3.0.16 is not up-to-date
+2017-01-17 09:36:42.407730 W | etcdserver: member fd32987dcd0511e0 has a higher version 3.1.0
+```
+
+#### 4. Repeat step 2 to step 3 for all other members
+
+#### 5. Finish
+
+When all members are upgraded, the cluster will report  upgrading to 3.1 successfully:
```
Comment (heyitsanthony) [inline Documentation/upgrades/upgrade_3_1.md]: stray double-space; s/  / /

File: Documentation/upgrades/upgrade_3_1.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+```
+2017-01-17 09:34:18.352662 I | raft: raft.node: 1640829d9eea5cfb elected leader 1640829d9eea5cfb at term 5
+2017-01-17 09:34:18.359630 W | etcdserver: failed to reach the peerURL(http://localhost:2380) of member fd32987dcd0511e0 (Get http://localhost:2380/version: dial tcp 127.0.0.1:2380: getsockopt: connection refused)
+2017-01-17 09:34:18.359679 W | etcdserver: cannot get the version of member fd32987dcd0511e0 (Get http://localhost:2380/version: dial tcp 127.0.0.1:2380: getsockopt: connection refused)
+2017-01-17 09:34:18.548116 W | rafthttp: lost the TCP streaming connection with peer fd32987dcd0511e0 (stream Message writer)
+2017-01-17 09:34:19.147816 W | rafthttp: lost the TCP streaming connection with peer fd32987dcd0511e0 (stream MsgApp v2 writer)
+2017-01-17 09:34:34.364907 W | etcdserver: failed to reach the peerURL(http://localhost:2380) of member fd32987dcd0511e0 (Get http://localhost:2380/version: dial tcp 127.0.0.1:2380: getsockopt: connection refused)
+```
+
+It’s a good idea at this point to  [backup the etcd data](../op-guide/maintenance.md#snapshot-backup) to provide a downgrade path should any problems occur:
```
Comment (heyitsanthony) [inline Documentation/upgrades/upgrade_3_1.md]: replace `’` with ascii `'`?

File: Documentation/upgrades/upgrade_3_1.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+Please [backup the data directory](../op-guide/maintenance.md#snapshot-backup) of all etcd members to make downgrading the cluster possible even after it has been completely upgraded.
+
+### Upgrade Procedure
+
+This example shows how to upgrade a 3-member v3.0 ectd cluster running on a local machine.
+
+#### 1. Check upgrade requirements.
+
+Is the the cluster healthy and running v.3.0.x?
```
Comment (heyitsanthony) [inline Documentation/upgrades/upgrade_3_1.md]: s/v.3.0.x/v3.0.x?

File: Documentation/upgrades/upgrade_3_1.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+If all members have been upgraded to v3.1, the cluster will be upgraded to v3.1, and downgrade from this completed state is **not possible**. If any single member is still v3.0, however, the cluster and its operations remains “v3.0”, and it is possible from this mixed cluster state to return to using a v3.0 etcd binary on all members.
+
+Please [backup the data directory](../op-guide/maintenance.md#snapshot-backup) of all etcd members to make downgrading the cluster possible even after it has been completely upgraded.
+
+### Upgrade Procedure
+
+This example shows how to upgrade a 3-member v3.0 ectd cluster running on a local machine.
+
+#### 1. Check upgrade requirements.
```
Comment (heyitsanthony) [inline Documentation/upgrades/upgrade_3_1.md]: s/requirements./requirements/?

File: Documentation/upgrades/upgrade_3_1.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+For a much larger total data size, 100MB or more , this one-time process might take even more time. Administrators of very large etcd clusters of this magnitude can feel free to contact the [etcd team][etcd-contact] before upgrading, and we’ll be happy to provide advice on the procedure.
+
+#### Downgrade
+
+If all members have been upgraded to v3.1, the cluster will be upgraded to v3.1, and downgrade from this completed state is **not possible**. If any single member is still v3.0, however, the cluster and its operations remains “v3.0”, and it is possible from this mixed cluster state to return to using a v3.0 etcd binary on all members.
+
+Please [backup the data directory](../op-guide/maintenance.md#snapshot-backup) of all etcd members to make downgrading the cluster possible even after it has been completely upgraded.
+
+### Upgrade Procedure
```
Comment (heyitsanthony) [inline Documentation/upgrades/upgrade_3_1.md]: s/Procedure/procedure to conform with heading style

File: Documentation/upgrades/upgrade_3_1.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+Before beginning, [backup the etcd data](../op-guide/maintenance.md#snapshot-backup). Should something go wrong with the upgrade, it is possible to use this backup to [downgrade](#downgrade) back to existing etcd version.
+
+#### Mixed Versions
+
+While upgrading, an etcd cluster supports mixed versions of etcd members, and operates with the protocol of the lowest common version. The cluster is only considered upgraded once all of its members are upgraded to version 3.1. Internally, etcd members negotiate with each other to determine the overall cluster version, which controls the reported version and the supported features.
+
+#### Limitations
+
+Note: If you only have data in v3 backend, you are not subject to this limitation.
```
Comment (heyitsanthony) [inline Documentation/upgrades/upgrade_3_1.md]: eschew you
`Note: If the cluster only has v3 data and no v2 data, it is not subject to this limitation.`

File: Documentation/upgrades/upgrade_3_1.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+If you have large v2 data, it might take up to 2 minutes for the newly upgraded member to catch up with the existing cluster when the total data size is larger than 50MB. Check the size of a recent  snapshot to estimate  the total data size. In other words, it is safest to wait for 2 minutes between upgrading each member.
```
Comment (heyitsanthony) [inline Documentation/upgrades/upgrade_3_1.md]: eschew you
`If the cluster is serving a v2 data set larger than 50MB, each newly upgraded member may take up to two minutes to catch up with the existing cluster.`

stray double space-- s/estimate  the/estimate the; s/recent  snapshot/ recent snapshot/

File: Documentation/upgrades/upgrade_3_1.md [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+To upgrade an existing etcd deployment to 3.1, the running cluster must be 3.0 or greater. If it's before 3.0, please upgrade to [3.0](https://github.com/coreos/etcd/releases/tag/v3.0.16) before upgrading to 3.1.
+
+Also, to ensure a smooth rolling upgrade, the running cluster must be healthy. You can check the health of the cluster by using the `etcdctl endpoint health` command.
+
+#### Preparation
+
+Before upgrading etcd, always test the services relying on etcd in a staging environment before deploying the upgrade to the production environment.
+
+Before beginning, [backup the etcd data](../op-guide/maintenance.md#snapshot-backup). Should something go wrong with the upgrade, it is possible to use this backup to [downgrade](#downgrade) back to existing etcd version.
```
Comment (heyitsanthony) [inline Documentation/upgrades/upgrade_3_1.md]: should there be some warning about v2 upgrades? snapshot won't pull that data

File: Documentation/upgrades/upgrade_3_1.md [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+Before [starting an upgrade](#upgrade-procedure), read through the rest of this guide to prepare.
+
+### Upgrade Checklists
+
+#### Upgrade Requirements
+
+To upgrade an existing etcd deployment to 3.1, the running cluster must be 3.0 or greater. If it's before 3.0, please upgrade to [3.0](https://github.com/coreos/etcd/releases/tag/v3.0.16) before upgrading to 3.1.
+
+Also, to ensure a smooth rolling upgrade, the running cluster must be healthy. You can check the health of the cluster by using the `etcdctl endpoint health` command.
```
Comment (heyitsanthony) [inline Documentation/upgrades/upgrade_3_1.md]: eschew you
`Check the health of the cluster by using the `etcdctl endpoint health` command before proceeding.`

File: Documentation/upgrades/upgrade_3_1.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+In the general case, upgrading from etcd 3.0 to 3.1 can be a zero-downtime, rolling upgrade:
+ - one by one, stop the etcd v3.0 processes and replace them with etcd v3.1 processes
+ - after running all v3.1 processes, new features in v3.1 are available to the cluster
+
+Before [starting an upgrade](#upgrade-procedure), read through the rest of this guide to prepare.
+
+### Upgrade Checklists
+
+#### Upgrade Requirements
```
Comment (heyitsanthony) [inline Documentation/upgrades/upgrade_3_1.md]: s/Requirements/requirements to conform with heading style

File: Documentation/upgrades/upgrade_3_1.md [hunk 13]
```diff
@@ -0,0 +1,123 @@
+## Upgrade etcd from 3.0 to 3.1
+
+In the general case, upgrading from etcd 3.0 to 3.1 can be a zero-downtime, rolling upgrade:
+ - one by one, stop the etcd v3.0 processes and replace them with etcd v3.1 processes
+ - after running all v3.1 processes, new features in v3.1 are available to the cluster
+
+Before [starting an upgrade](#upgrade-procedure), read through the rest of this guide to prepare.
+
+### Upgrade Checklists
```
Comment (heyitsanthony) [inline Documentation/upgrades/upgrade_3_1.md]: s/Checklists/checklists to conform with heading style

**Comment by gyuho** [comment]
@heyitsanthony All fixed. PTAL. Thanks.

File: Documentation/upgrades/upgrade_3_1.md [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+Please [backup the data directory](../op-guide/maintenance.md#snapshot-backup) of all etcd members to make downgrading the cluster possible even after it has been completely upgraded.
+
+### Upgrade procedure
+
+This example shows how to upgrade a 3-member v3.0 ectd cluster running on a local machine.
+
+#### 1. Check upgrade requirements
+
+Is the the cluster healthy and running v3.0.x?
```
Comment (heyitsanthony) [inline Documentation/upgrades/upgrade_3_1.md]: s/the the/the

**Comment by heyitsanthony** [comment]
lgtm after fixing the the the

**Pull request merged**